File: CastVideos-objc/Classes/MediaListModel.m
// Copyright 2022 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "MediaListModel.h"

#import <GoogleCast/GoogleCast.h>

#import "MediaItem.h"

NSString *const kMediaKeyPosterURL = @"posterUrl";
NSString *const kMediaKeyDescription = @"description";

static NSString *const kKeyCategories = @"categories";
static NSString *const kKeyHlsBaseURL = @"hls";
static NSString *const kKeyImagesBaseURL = @"images";
static NSString *const kKeyTracksBaseURL = @"tracks";
static NSString *const kKeySources = @"sources";
static NSString *const kKeyVideos = @"videos";
static NSString *const kKeyArtist = @"artist";
static NSString *const kKeyBaseURL = @"baseUrl";
static NSString *const kKeyContentID = @"contentId";
static NSString *const kKeyDescription = @"description";
static NSString *const kKeyID = @"id";
static NSString *const kKeyImageURL = @"image-480x270";
static NSString *const kKeyItems = @"items";
static NSString *const kKeyLanguage = @"language";
static NSString *const kKeyMimeType = @"mime";
static NSString *const kKeyName = @"name";
static NSString *const kKeyPosterURL = @"image-780x1200";
static NSString *const kKeyStreamType = @"streamType";
static NSString *const kKeyStudio = @"studio";
static NSString *const kKeySubtitle = @"subtitle";
static NSString *const kKeySubtype = @"subtype";
static NSString *const kKeyTitle = @"title";
static NSString *const kKeyTracks = @"tracks";
static NSString *const kKeyType = @"type";
static NSString *const kKeyURL = @"url";
static NSString *const kKeyDuration = @"duration";

static NSString *const kDefaultVideoMimeType = @"application/x-mpegurl";
static NSString *const kDefaultTrackMimeType = @"text/vtt";

static NSString *const kTypeAudio = @"audio";
static NSString *const kTypePhoto = @"photos";
static NSString *const kTypeVideo = @"videos";

static NSString *const kTypeLive = @"live";

static const NSInteger kThumbnailWidth = 480;
static const NSInteger kThumbnailHeight = 720;

static const NSInteger kPosterWidth = 780;
static const NSInteger kPosterHeight = 1200;

@interface MediaListModel () {
  NSURLSession *_session;
  NSURLSessionDataTask *_dataTask;
  GCKMediaTextTrackStyle *_trackStyle;
  MediaItem *_rootItem;
}

@property(nonatomic, readwrite) NSString *title;
@property(nonatomic, assign, readwrite) BOOL loaded;

@end

@implementation MediaListModel {
  /** Storage for the list of Media objects. */
  NSArray *_medias;
}

- (instancetype)init {
  if (self = [super init]) {
    _trackStyle = [GCKMediaTextTrackStyle createDefault];
    _session = [NSURLSession sessionWithConfiguration: [NSURLSessionConfiguration defaultSessionConfiguration]];

  }
  return self;
}

- (void)loadFromURL:(NSURL *)url {
  _rootItem = nil;

  GCKLog(@"loading media list from URL %@", url);

  if (_dataTask != nil) {
    [_dataTask cancel];
  }

  __weak MediaListModel *weakSelf = self;
  _dataTask = [_session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
    if (error != nil) {
      GCKLog(@"httpRequest failed with %@", error);
      dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf.delegate mediaListModel:weakSelf didFailToLoadWithError:error];
      });
      return;
    }

    long responseStatus = ((NSHTTPURLResponse *)response).statusCode;

    GCKLog(@"httpRequest completed with %ld", responseStatus);
    if (responseStatus == 200) {
      NSError *error;
      NSDictionary *jsonData = [NSJSONSerialization JSONObjectWithData:data
                                                               options:kNilOptions
                                                                 error:&error];
      self->_rootItem = [weakSelf decodeMediaTreeFromJSON:jsonData];
      weakSelf.loaded = YES;
      dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf.delegate mediaListModelDidLoad:weakSelf];
      });
    } else {
      NSError *error = [[NSError alloc] initWithDomain:@"HTTP" code:responseStatus userInfo:nil];
      dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf.delegate mediaListModel:weakSelf didFailToLoadWithError:error];
      });
    }
  }];

  if (_dataTask != nil) {
    [_dataTask resume];
  }
}

- (MediaItem *)rootItem {
  return _rootItem;
}

#pragma mark - JSON decoding

- (MediaItem *)decodeMediaTreeFromJSON:(NSDictionary *)json {
  MediaItem *rootItem = [[MediaItem alloc] initWithTitle:nil imageURL:nil parent:nil];

  NSArray *categories = [json gck_arrayForKey:kKeyCategories];
  for (NSDictionary *categoryElement in categories) {
    if (![categoryElement isKindOfClass:[NSDictionary class]]) continue;

    NSDictionary *category = (NSDictionary *)categoryElement;

    NSArray *mediaList = [category gck_arrayForKey:kKeyVideos];
    if (mediaList && [mediaList isKindOfClass:[NSArray class]]) {
      self.title = [category gck_stringForKey:kKeyName];
      // Pick the MP4 files only
      NSString *videosBaseURLString = [category gck_stringForKey:kKeyHlsBaseURL];
      NSURL *videosBaseURL = [NSURL URLWithString:videosBaseURLString];
      NSString *imagesBaseURLString = [category gck_stringForKey:kKeyImagesBaseURL];
      NSURL *imagesBaseURL = [NSURL URLWithString:imagesBaseURLString];
      NSString *tracksBaseURLString = [category gck_stringForKey:kKeyTracksBaseURL];
      NSURL *tracksBaseURL = [NSURL URLWithString:tracksBaseURLString];
      [self decodeItemListFromArray:mediaList
                           intoItem:rootItem
                        videoFormat:kKeyHlsBaseURL
                      videosBaseURL:videosBaseURL
                      imagesBaseURL:imagesBaseURL
                      tracksBaseURL:tracksBaseURL];
      break;
    }
  }

  return rootItem;
}

- (NSURL *)buildURLWithString:(NSString *)string baseURL:(NSURL *)baseURL {
  if (!string) return nil;

  if ([string hasPrefix:@"http://"] || [string hasPrefix:@"https://"]) {
    return [NSURL URLWithString:string];
  } else {
    return [NSURL URLWithString:string relativeToURL:baseURL];
  }
}

- (void)decodeItemListFromArray:(NSArray *)array
                       intoItem:(MediaItem *)item
                    videoFormat:(NSString *)videoFormat
                  videosBaseURL:(NSURL *)videosBaseURL
                  imagesBaseURL:(NSURL *)imagesBaseURL
                  tracksBaseURL:(NSURL *)tracksBaseURL {
  for (NSDictionary *element in array) {
    if (![element isKindOfClass:[NSDictionary class]]) continue;

    NSDictionary *dict = (NSDictionary *)element;

    NSString *title = [dict gck_stringForKey:kKeyTitle];

    GCKMediaMetadata *metadata =
        [[GCKMediaMetadata alloc] initWithMetadataType:GCKMediaMetadataTypeMovie];
    [metadata setString:title forKey:kGCKMetadataKeyTitle];

    NSString *mimeType = nil;
    NSURL *url = nil;
    NSArray *sources = [dict gck_arrayForKey:kKeySources];
    for (id sourceElement in sources) {
      if (![sourceElement isKindOfClass:[NSDictionary class]]) continue;

      NSDictionary *sourceDict = (NSDictionary *)sourceElement;

      NSString *type = [sourceDict gck_stringForKey:kKeyType];
      if ([type isEqualToString:videoFormat]) {
        mimeType = [sourceDict gck_stringForKey:kKeyMimeType];
        NSString *urlText = [sourceDict gck_stringForKey:kKeyURL];
        url = [self buildURLWithString:urlText baseURL:videosBaseURL];
        break;
      }
    }

    NSString *imageURLString = [dict gck_stringForKey:kKeyImageURL];
    NSURL *imageURL = [self buildURLWithString:imageURLString baseURL:imagesBaseURL];
    if (imageURL) {
      [metadata addImage:[[GCKImage alloc] initWithURL:imageURL
                                                 width:kThumbnailWidth
                                                height:kThumbnailHeight]];
    }

    NSString *posterURLText = [dict gck_stringForKey:kKeyPosterURL];
    NSURL *posterURL = [self buildURLWithString:posterURLText baseURL:imagesBaseURL];
    if (posterURL) {
      [metadata setString:posterURL.absoluteString forKey:kMediaKeyPosterURL];
      [metadata addImage:[[GCKImage alloc] initWithURL:posterURL
                                                 width:kPosterWidth
                                                height:kPosterHeight]];
    }

    NSString *description = [dict gck_stringForKey:kKeySubtitle];
    if (description) {
      [metadata setString:description forKey:kMediaKeyDescription];
    }

    NSMutableArray *mediaTracks = nil;

    NSString *studio = [dict gck_stringForKey:kKeyStudio];
    [metadata setString:studio forKey:kGCKMetadataKeyStudio];

    NSInteger duration = [dict gck_integerForKey:kKeyDuration];

    mediaTracks = [[NSMutableArray alloc] init];

    NSArray *tracks = [dict gck_arrayForKey:kKeyTracks];
    for (id trackElement in tracks) {
      if (![trackElement isKindOfClass:[NSDictionary class]]) continue;

      NSDictionary *trackDict = (NSDictionary *)trackElement;

      NSInteger identifier = [trackDict gck_integerForKey:kKeyID];
      NSString *name = [trackDict gck_stringForKey:kKeyName];
      NSString *typeString = [trackDict gck_stringForKey:kKeyType];
      NSString *subtypeString = [trackDict gck_stringForKey:kKeySubtype];
      NSString *contentID = [trackDict gck_stringForKey:kKeyContentID];
      NSString *language = [trackDict gck_stringForKey:kKeyLanguage];

      NSURL *url = [self buildURLWithString:contentID baseURL:tracksBaseURL];

      GCKMediaTrack *mediaTrack =
          [[GCKMediaTrack alloc] initWithIdentifier:identifier
                                  contentIdentifier:url.absoluteString
                                        contentType:kDefaultTrackMimeType
                                               type:[self trackTypeFrom:typeString]
                                        textSubtype:[self textTrackSubtypeFrom:subtypeString]
                                               name:name
                                       languageCode:language
                                         customData:nil];
      [mediaTracks addObject:mediaTrack];
    }
    if (mediaTracks.count == 0) {
      mediaTracks = nil;
    }

    // Define information about the media item.
    GCKMediaInformationBuilder *mediaInfoBuilder =
        [[GCKMediaInformationBuilder alloc] initWithContentURL:url];
    mediaInfoBuilder.streamDuration = duration;
    mediaInfoBuilder.streamType = GCKMediaStreamTypeBuffered;
    mediaInfoBuilder.contentType = mimeType;
    mediaInfoBuilder.metadata = metadata;
    mediaInfoBuilder.mediaTracks = mediaTracks;
    mediaInfoBuilder.textTrackStyle = _trackStyle;
    GCKMediaInformation *mediaInfo = [mediaInfoBuilder build];

    MediaItem *childItem = [[MediaItem alloc] initWithMediaInformation:mediaInfo parent:item];
    [item.children addObject:childItem];
  }
}

- (GCKMediaTrackType)trackTypeFrom:(NSString *)string {
  if ([string isEqualToString:@"audio"]) {
    return GCKMediaTrackTypeAudio;
  }
  if ([string isEqualToString:@"text"]) {
    return GCKMediaTrackTypeText;
  }
  if ([string isEqualToString:@"video"]) {
    return GCKMediaTrackTypeVideo;
  }
  return GCKMediaTrackTypeUnknown;
}

- (GCKMediaTextTrackSubtype)textTrackSubtypeFrom:(NSString *)string {
  if ([string isEqualToString:@"captions"]) {
    return GCKMediaTextTrackSubtypeCaptions;
  }
  if ([string isEqualToString:@"chapters"]) {
    return GCKMediaTextTrackSubtypeChapters;
  }
  if ([string isEqualToString:@"descriptions"]) {
    return GCKMediaTextTrackSubtypeDescriptions;
  }
  if ([string isEqualToString:@"metadata"]) {
    return GCKMediaTextTrackSubtypeMetadata;
  }
  if ([string isEqualToString:@"subtitles"]) {
    return GCKMediaTextTrackSubtypeSubtitles;
  }

  return GCKMediaTextTrackSubtypeUnknown;
}

@end

File: CastVideos-swift/Classes/MediaListModel.swift
// Copyright 2018 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
import GoogleCast

/** A key for the URL of the media item's poster (large image). */
let kMediaKeyPosterURL = "posterUrl"
/** A key for the media item's extended description. */
let kMediaKeyDescription = "description"
let kKeyCategories = "categories"
let kKeyMP4BaseURL = "mp4"
let kKeyImagesBaseURL = "images"
let kKeyTracksBaseURL = "tracks"
let kKeySources = "sources"
let kKeyVideos = "videos"
let kKeyArtist = "artist"
let kKeyBaseURL = "baseUrl"
let kKeyContentID = "contentId"
let kKeyDescription = "description"
let kKeyID = "id"
let kKeyImageURL = "image-480x270"
let kKeyItems = "items"
let kKeyLanguage = "language"
let kKeyMimeType = "mime"
let kKeyName = "name"
let kKeyPosterURL = "image-780x1200"
let kKeyStreamType = "streamType"
let kKeyStudio = "studio"
let kKeySubtitle = "subtitle"
let kKeySubtype = "subtype"
let kKeyTitle = "title"
let kKeyTracks = "tracks"
let kKeyType = "type"
let kKeyURL = "url"
let kKeyDuration = "duration"
let kDefaultVideoMimeType = "video/mp4"
let kDefaultTrackMimeType = "text/vtt"
let kTypeAudio = "audio"
let kTypePhoto = "photos"
let kTypeVideo = "videos"
let kTypeLive = "live"
let kThumbnailWidth = 480
let kThumbnailHeight = 720
let kPosterWidth = 780
let kPosterHeight = 1200

/**
 * The delegate protocol for receiving notifications from the model.
 */
protocol MediaListModelDelegate: NSObjectProtocol {
  /**
   * Called when the media list has loaded.
   *
   * @param list The media list.
   */
  func mediaListModelDidLoad(_ list: MediaListModel)

  /**
   * Called when the media list has failed to load.
   *
   * @param list The media list.
   * @param error The error.
   */
  func mediaListModel(_ list: MediaListModel, didFailToLoadWithError error: Error?)
}

/**
 * An object representing a hierarchy of media items.
 */
class MediaListModel: NSObject, NSURLConnectionDelegate, NSURLConnectionDataDelegate {
  private var request: URLRequest!
  private var connection: NSURLConnection!
  private var responseData: Data!
  private var responseStatus: Int = 0
  private var trackStyle: GCKMediaTextTrackStyle!
  /* The root item (top-level group). */
  fileprivate(set) var rootItem: MediaItem!

  /** A delegate for receiving notifications from the model. */
  weak var delegate: MediaListModelDelegate?
  /** A flag indicating whether the model has been loaded. */
  fileprivate(set) var isLoaded: Bool = false
  /** The title of the media list. */
  fileprivate(set) var title: String = ""

  /** Storage for the list of Media objects. */
  var medias = [Any]()

  /**
   * Begins loading the model from the given URL. The delegate will be messaged
   * when the load
   * completes or fails.
   *
   * @param url The URL of the JSON file describing the media hierarchy.
   */
  func load(from url: URL) {
    rootItem = nil
    request = URLRequest(url: url)
    connection = NSURLConnection(request: request, delegate: self)
    responseData = nil
    connection.start()
    GCKLogger.sharedInstance().delegate?.logMessage?("loading media list from URL \(url)", at: .debug, fromFunction: #function, location: "MediaListModel.class")
  }

  override init() {
    super.init()

    trackStyle = GCKMediaTextTrackStyle.createDefault()
  }

  func cancelLoad() {
    if request != nil {
      connection.cancel()
      request = nil
      connection = nil
      responseData = nil
    }
  }

  // MARK: - NSURLConnectionDelegate

  func connection(_: NSURLConnection, didFailWithError error: Error) {
    request = nil
    responseData = nil
    connection = nil
    GCKLogger.sharedInstance().delegate?.logMessage?("httpRequest failed with \(error)", at: .debug, fromFunction: #function, location: "MediaListModel.class")
    delegate?.mediaListModel(self, didFailToLoadWithError: error)
  }

  // MARK: - NSURLConnectionDataDelegate

  internal func connection(_: NSURLConnection, didReceive response: URLResponse) {
    if let response = response as? HTTPURLResponse {
      responseStatus = response.statusCode
    }
  }

  func connection(_: NSURLConnection, didReceive data: Data) {
    if responseData == nil {
      responseData = Data()
    }
    responseData.append(data)
  }

  func connectionDidFinishLoading(_: NSURLConnection) {
    GCKLogger.sharedInstance().delegate?.logMessage?("httpRequest completed with \(responseStatus)", at: .debug, fromFunction: #function, location: "MediaListModel.class")

    if responseStatus == 200 {
      let jsonData = (try? JSONSerialization.jsonObject(with: responseData,
                                                        options: .mutableContainers)) as? NSDictionary
      rootItem = decodeMediaTree(fromJSON: jsonData!)
      isLoaded = true
      delegate?.mediaListModelDidLoad(self)
    } else {
      let error = NSError(domain: "HTTP", code: responseStatus, userInfo: nil)
      delegate?.mediaListModel(self, didFailToLoadWithError: error)
    }
  }

  // MARK: - JSON decoding

  func decodeMediaTree(fromJSON json: NSDictionary) -> MediaItem {
    let rootItem = MediaItem(title: nil, imageURL: nil, parent: nil)
    let categories = json.gck_array(forKey: kKeyCategories)!
    for categoryElement in categories {
      if !(categoryElement is [AnyHashable: Any]) {
        continue
      }
      let category = (categoryElement as? NSDictionary)
      let mediaList = category?.gck_array(forKey: kKeyVideos)
      if mediaList != nil {
        title = (category?.gck_string(forKey: kKeyName))!
        // Pick the MP4 files only
        let videosBaseURLString: String? = category?.gck_string(forKey: kKeyMP4BaseURL)
        let videosBaseURL = URL(string: videosBaseURLString!)
        let imagesBaseURLString: String? = category?.gck_string(forKey: kKeyImagesBaseURL)
        let imagesBaseURL = URL(string: imagesBaseURLString!)
        let tracksBaseURLString: String? = category?.gck_string(forKey: kKeyTracksBaseURL)
        let tracksBaseURL = URL(string: tracksBaseURLString!)
        decodeItemList(fromArray: mediaList!, into: rootItem, videoFormat: kKeyMP4BaseURL,
                       videosBaseURL: videosBaseURL!, imagesBaseURL: imagesBaseURL!, tracksBaseURL: tracksBaseURL!)
        break
      }
    }
    return rootItem
  }

  func buildURL(with string: String?, baseURL: URL) -> URL? {
    guard let string = string else { return nil }
    if string.hasPrefix("http://") || string.hasPrefix("https://") {
      return URL(string: string)!
    } else {
      return URL(string: string, relativeTo: baseURL)!
    }
  }

  func decodeItemList(fromArray array: [Any], into item: MediaItem, videoFormat: String,
                      videosBaseURL: URL, imagesBaseURL: URL, tracksBaseURL: URL) {
    for element in array {
      if let dict = element as? NSDictionary {
        let metadata = GCKMediaMetadata(metadataType: .movie)
        if let title = dict.gck_string(forKey: kKeyTitle) {
          metadata.setString(title, forKey: kGCKMetadataKeyTitle)
        }
        var mimeType: String?
        var url: URL?
        let sources: [Any] = dict.gck_array(forKey: kKeySources) ?? []
        for sourceElement in sources {
          if let sourceDict = sourceElement as? NSDictionary {
            let type: String? = sourceDict.gck_string(forKey: kKeyType)
            if type == videoFormat {
              mimeType = sourceDict.gck_string(forKey: kKeyMimeType)
              let urlText: String? = sourceDict.gck_string(forKey: kKeyURL)
              url = buildURL(with: urlText, baseURL: videosBaseURL)
              break
            }
          }
        }
        let imageURLString: String? = dict.gck_string(forKey: kKeyImageURL)
        if let imageURL = self.buildURL(with: imageURLString, baseURL: imagesBaseURL) {
          metadata.addImage(GCKImage(url: imageURL, width: kThumbnailWidth, height: kThumbnailHeight))
        }
        let posterURLText: String? = dict.gck_string(forKey: kKeyPosterURL)
        if let posterURL = self.buildURL(with: posterURLText, baseURL: imagesBaseURL) {
          metadata.setString(posterURL.absoluteString, forKey: kMediaKeyPosterURL)
          metadata.addImage(GCKImage(url: posterURL, width: kPosterWidth, height: kPosterHeight))
        }
        if let description = dict.gck_string(forKey: kKeySubtitle) {
          metadata.setString(description, forKey: kMediaKeyDescription)
        }
        var mediaTracks: [GCKMediaTrack]?
        if let studio = dict.gck_string(forKey: kKeyStudio) {
          metadata.setString(studio, forKey: kGCKMetadataKeyStudio)
        }
        let duration: Int? = dict.gck_integer(forKey: kKeyDuration)
        mediaTracks = [GCKMediaTrack]()
        if let tracks = dict.gck_array(forKey: kKeyTracks) {
          for trackElement: Any in tracks {
            guard let trackDict = trackElement as? NSDictionary else { continue }
            let identifier = trackDict.gck_integer(forKey: kKeyID)
            let name: String? = trackDict.gck_string(forKey: kKeyName)
            let typeString: String? = trackDict.gck_string(forKey: kKeyType)
            let subtypeString: String? = trackDict.gck_string(forKey: kKeySubtype)
            let contentID: String? = trackDict.gck_string(forKey: kKeyContentID)
            let language: String? = trackDict.gck_string(forKey: kKeyLanguage)
            let url = buildURL(with: contentID, baseURL: tracksBaseURL)
            let mediaTrack = GCKMediaTrack(identifier: identifier, contentIdentifier: url?.absoluteString,
                                           contentType: kDefaultTrackMimeType, type: trackType(from: typeString!),
                                           textSubtype: textTrackSubtype(from: subtypeString!),
                                           name: name, languageCode: language, customData: nil)
            mediaTracks?.append(mediaTrack)
          }
        }

        if mediaTracks?.count == 0 {
          mediaTracks = nil
        }

        let mediaInfoBuilder = GCKMediaInformationBuilder(contentURL: url!)
        // TODO: Remove contentID when sample receiver supports using contentURL
        mediaInfoBuilder.contentID = url!.absoluteString
        mediaInfoBuilder.streamType = .buffered
        mediaInfoBuilder.streamDuration = TimeInterval(duration!)
        mediaInfoBuilder.contentType = mimeType!
        mediaInfoBuilder.metadata = metadata
        mediaInfoBuilder.mediaTracks = mediaTracks
        mediaInfoBuilder.textTrackStyle = trackStyle

        let mediaInfo = mediaInfoBuilder.build()
        let childItem = MediaItem(mediaInformation: mediaInfo, parent: item)
        item.children.append(childItem)
      }
    }
  }

  func trackType(from string: String) -> GCKMediaTrackType {
    if string == "audio" {
      return .audio
    }
    if string == "text" {
      return .text
    }
    if string == "video" {
      return .video
    }
    return .unknown
  }

  func textTrackSubtype(from string: String) -> GCKMediaTextTrackSubtype {
    if string == "captions" {
      return .captions
    }
    if string == "chapters" {
      return .chapters
    }
    if string == "descriptions" {
      return .descriptions
    }
    if string == "metadata" {
      return .metadata
    }
    if string == "subtitles" {
      return .subtitles
    }
    return .unknown
  }
}
